cpp/include/cudf/table/experimental/row_operators.cuh

```diff
@@ -824,8 +824,15 @@ class two_table_comparator {
 }  // namespace lexicographic
 
 namespace hash {
+
+template <template <typename> class hash_function, typename Nullate>
+class device_row_hasher;
+
+template <template <template <typename> class hash_function, typename> class DeviceRowHasher =
```

Comment:
One thought. Instead of templating the `row_hasher` class itself, would it be cleaner to template just the `device_hasher` method? The upside is that instead of instantiating the entire class multiple times, we just instantiate that method multiple times, and the feature is more effectively localized. From a usage perspective it means that any call to `device_hasher` that doesn't use a custom adapter now becomes a bit more verbose because you have to do `hasher.device_hasher<adapter<hash_function>>` instead of just `hasher.device_hasher<hash_function>`, but it makes the code a bit cleaner when you _do_ have a custom adapter because all the templating is isolated to just the function call rather than the constructor. I could go either way, but wanted to consider the option now before we finalize.

Reply:
This was a really great idea. It simplified the code a lot. (Originally I thought there was a reason why this couldn't work, which led me to template the entire `row_hasher` class. Perhaps my design changed along the way and made this approach viable.) Thank you! See ef0ad3bc2c6ea0d3093ec6561a00befbdd434c41.